dace/transformation/dataflow/mapreduce.py

```diff
@@ -133,7 +133,7 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
 
         # Add initialization state as necessary
         if not self.no_init and reduce_node.identity is not None:
-            init_state = sdfg.add_state_before(graph)
+            init_state = graph.parent_graph.add_state_before(graph)
```

Comment:
I mean this comes with making the SDFG hierarchical, but I suppose a method wrapper for `add_state_before` can be used on the SDFG that traverses the SDFG to first find `graph` and their parent and then insert a state there.
However, I do have to say that I see that method as significantly more error prone than _not_ providing it, which would cause a catastrophic failure that immediately exposes the issue and allows for (an easy) adaptation, as opposed to possibly silently creating entirely wrong SDFGs.
We can discuss alternatives if you have ideas.